lang/src/program_data.rs

```diff
@@ -0,0 +1,143 @@
+use crate::error::ErrorCode;
+use crate::*;
+use solana_program::bpf_loader_upgradeable;
+use solana_program::entrypoint::ProgramResult;
+use solana_program::instruction::AccountMeta;
+use solana_program::program_error::ProgramError;
+use solana_program::pubkey::Pubkey;
+use solana_program::{account_info::AccountInfo, bpf_loader_upgradeable::UpgradeableLoaderState};
+use std::ops::Deref;
+
+#[derive(Debug, Clone, AnchorSerialize, AnchorDeserialize)]
+pub struct ProgramDataInner {
+    slot: u64,
+    upgrade_authority_address: Option<Pubkey>,
+}
+
+impl ProgramDataInner {
+    pub fn slot(&self) -> &u64 {
+        &self.slot
+    }
+
+    pub fn upgrade_authority_address(&self) -> &Option<Pubkey> {
+        &self.upgrade_authority_address
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct ProgramData<'info> {
```

Comment:
What do you think about making a wrapper type for `UpgradeableLoaderState` that implements the required traits? Then we could use `Account<UpgradeableLoaderState>` vs `ProgramData`.

Reply:
sg,Ill try.

 although I'd prefer to have an `Account<ProgramData>` and an `Account<Program>` instead of `Account<UpgradeableLoaderState>`. or all three of them. cause the first two are useful for constraints. You can't match easily on an enum.

Reply:
hmm actually idk if `Account<Program>` is a good idea. We already have `Program<'info, PROGRAM>` which also checks that we have the correct program. We could just add the data of the program (=programdata_address) to that struct as an option. It would be Some(..) if the given program is owned by the upgradable loader.

`Account<ProgramData>` and `Account<UpgradableLoaderState>` still sg to me tho

Reply:
I've added ProgramData and UgpradableLoaderState and they work, just gotta decide what to do with Program

Reply:
Let's add `Option<Pubkey>` to `Program<'info, T>` for the programdata_address. I'm not a fan of `Account<Program>` since there are no address checks done.

Reply:
yup, agree.

can merge this pr and do that in a separate issue tho. theyre not connected.

Reply:
already created i some days ago https://github.com/project-serum/anchor/issues/1094

Reply:
Sounds good.